pkg/status/workstatus.go

```diff
@@ -0,0 +1,178 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+const workStatusIdentificationIndexKey = "workStatusIdentificationIndex"
+
+type workStatus struct {
+	workStatusRef
+	status map[string]interface{}
+}
+
+func (c *Controller) syncWorkStatus(ctx context.Context, ref workStatusRef) error {
+	workStatus := &workStatus{
+		workStatusRef: ref, //readonly
+		status:        nil,
+	}
+
+	obj, err := c.workStatusLister.ByNamespace(ref.wecName).Get(ref.name)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return fmt.Errorf("failed to get workstatus (%v): %w", ref, err)
+		}
+	} else {
+		status, err := util.GetWorkStatusStatus(obj)
+		if err != nil {
+			return fmt.Errorf("failed to get status from workstatus (%v): %w", ref, err)
+		}
+
+		workStatus.status = status // might be nil
+	}
+
+	combinedStatusSet := c.combinedStatusResolver.NoteWorkStatus(workStatus) // nil .status is equivalent to deleted
+	for combinedStatus := range combinedStatusSet {
+		c.workqueue.AddAfter(combinedStatusRef(combinedStatus.ObjectName.AsNamespacedName().String()), queueingDelay)
+	}
+
+	// only process workstatues with the label for single reported status
+	statusLabelVal, ok := obj.(metav1.Object).GetLabels()[util.BindingPolicyLabelSingletonStatusKey]
+	if !ok {
+		return nil
+	}
+
+	// remove the status if singleton status label value is unset
+	if statusLabelVal == util.BindingPolicyLabelSingletonStatusValueUnset {
+		emptyStatus := make(map[string]interface{})
+		return updateObjectStatus(ctx, &ref.sourceObjectIdentifier, emptyStatus,
+			c.listers, c.wdsDynClient)
+	}
+
+	status, err := util.GetWorkStatusStatus(obj)
+	if err != nil {
+		return err
+	}
+
+	if status == nil {
+		// when status is updated, a new event will trigger update
+		return nil
+	}
+
+	c.logger.Info("updating singleton status", "objectIdentifier", ref.sourceObjectIdentifier)
+	return updateObjectStatus(ctx, &ref.sourceObjectIdentifier, status, c.listers, c.wdsDynClient)
+}
+
+func updateObjectStatus(ctx context.Context, objectIdentifier *util.ObjectIdentifier, status map[string]interface{},
+	listers util.ConcurrentMap[schema.GroupVersionResource, cache.GenericLister], wdsDynClient dynamic.Interface) error {
+	logger := klog.FromContext(ctx)
+
+	gvr := objectIdentifier.GVR()
+	lister, found := listers.Get(gvr)
+	if !found {
+		logger.V(5).Info("could not find lister for gvr", "gvr", gvr)
+		return nil
+	}
+
+	obj, err := lister.ByNamespace(objectIdentifier.ObjectName.Namespace).Get(objectIdentifier.ObjectName.Name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.V(5).Info("could not find object", "objectIdentifier", objectIdentifier)
+			return nil
+		}
+
+		return fmt.Errorf("failed to get object (%v): %w", objectIdentifier, err)
+	}
+
+	unstrObj, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("object cannot be cast to *unstructured.Unstructured: object: %s", util.RefToRuntimeObj(obj))
```

Comment:
Is this a transient condition that is likely to go away when retried?

Reply:
This function will still return this error, but the caller will not retry (https://github.com/kubestellar/kubestellar/pull/2174#discussion_r1646316239)